File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(!head || !head->next) return head; 

        vector<ListNode*> result; 
        ListNode* temp = head; 

        while(temp){
            result.push_back(temp); 
            temp = temp->next; 
        }
        
        ListNode* connect = nullptr;  
        for(int i = 0; i < result.size()-1; i+=2){
            ListNode* save = result[i + 1]->next; 

            result[i + 1]->next = result[i];
            result[i]->next = save;

            if(connect){
                connect->next = result[i + 1]; 
            }

            connect = result[i]; 

            


             
            
           
            if(i == 0){
                head = result[i + 1]; 
            }
        }

        return head; 
        
    }
};

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)

        prev, curr = dummy, head 

        while curr and curr.next: #this is as we want a pair inorder to swap 
            #save some pointers first we need the next pair, and the next node to swap 
            nextPair = curr.next.next 
            toSwap = curr.next 

            #swap 
            curr.next = nextPair 
            toSwap.next = curr
            prev.next = toSwap #make the second node the head or the start of the pair

            #move the ptrs 
            prev = curr 
            curr = nextPair 
        
        return dummy.next 
        